### ISSUE TITLE
Softmax max exp_table lookup may include the sign bit

### ISSUE THREAD

**Issue by calad0i** [fastmachinelearning/hls4ml#1221]
https://github.com/fastmachinelearning/hls4ml/blob/0aa705b4dd83b2dddbbbd60f7fd006741027d1bb/hls4ml/templates/vivado/nnet_utils/nnet_activation.h#L144

In the current vivado/vitis implementation for softmax, the `exp_table` in the softmax function initializes the lookup table with the sign bit. For the stable implementation, as the inputs are almost always negative (`x-x_max`), Using MSB for sampling here causes some inefficiency (half of the table will never be accessed).

The proposed fix is as follows:
When using `stable` strategy, index the `exp_table` by `x_max - x` on an "inverse" `exp_table` of `exp(-x). With this, we are sure that the inputs to the `exp_table` are always non-negative, and may use `ap_ufixed`, where the MSB would make sense. 
Proposed fix is in included in [commit #b4eb0250](https://github.com/fastmachinelearning/hls4ml/pull/1191/commits/b4eb02504a1411f39cd674f412cc04569cffbfe7). This issue is just for record purpose.

Minimal example:

Add `std::cout << "exp_res[" << i << "](" << d_xi_xmax[i].to_float() << "->" << x << ") = " << exp_res[i].to_float() << std::endl;` after L256 of `hls4ml/templates/vivado/nnet_utils/nnet_activation.h`

```python
import keras
import numpy as np
from hls4ml.converters import convert_from_keras_model
from hls4ml.utils.config import config_from_keras_model

model = keras.Sequential([
    keras.layers.Input(shape=(5,)),
    keras.layers.Softmax(name='wtf')
])

dat = np.array([[0,1,2,3,4]], dtype=np.float32)*0.1

conf = {'Model': {'ReuseFactor': 1, 'Precision': 'ap_fixed<18,8>'}}
model_hls = convert_from_keras_model(model, hls_config=conf, output_dir='/tmp/tt', io_type='io_stream')

model_hls.write()
model_hls._compile()

model_hls.predict(dat)
```

The outputs as of the current codebase:
```
exp_res[0](-0.399414->1022) = 0.606445
exp_res[1](-0.299805->1022) = 0.606445
exp_res[2](-0.200195->1023) = 0.77832
exp_res[3](-0.0996094->1023) = 0.77832
exp_res[4](0->0) = 1
array([0.16113281, 0.16113281, 0.20703125, 0.20703125, 0.26660156],
      dtype=float32)
```

The outputs after the fix:
```
exp_res[0](0.399414->3) = 0.6875
exp_res[1](0.299805->2) = 0.77832
exp_res[2](0.200195->1) = 0.882812
exp_res[3](0.0996094->0) = 1
exp_res[4](0->0) = 1
array([0.16113281, 0.18261719, 0.20703125, 0.23535156, 0.23535156],
      dtype=float32)
```

Tested on (abused) the `test_conv1d` test to compute MAE:

Current:
```
[Vivado-io_parallel-latency]: MAE = 0.009399749017506839
[Vivado-io_stream-latency]: MAE = 0.00942513467296958
[Vivado-io_stream-resource]: MAE = 0.00947911877334118
[Vitis-io_stream-latency]: MAE = 0.009372118694335222
[Vivado-io_parallel-resourceMAE = ]: 0.009706129374355077
[Vitis-io_stream-resource]: MAE = 0.009281127255409955
[Vitis-io_parallel-resource]: MAE = 0.009532290701568127
[Vitis-io_parallel-latency]: MAE = 0.009174691569060088
```
After fix:
```
[Vivado-io_parallel-resource] - MAE = 0.0072945020921528336
[Vivado-io_parallel-latency] - MAE = 0.007336324921250343
[Vivado-io_stream-resource] - MAE = 0.007447532872110605
[Vivado-io_stream-latency] - MAE = 0.007666043180972338
[Vitis-io_stream-latency] - MAE = 0.007586525531858206
[Vitis-io_stream-resource] - MAE = 0.0073784718409180645
[Vitis-io_parallel-resource] - MAE = 0.007544468995928764
[Vitis-io_parallel-latency] - MAE = 0.007346910997480154
```

However, argmax results after the fix does not always be worse than the current version. Though, softmax should never be implemented on FPGAs for the purpose of classification anyway... Or one can use the argmax instead.
